BootloaderCommonPkg/Library/ShellLib/CmdLog.c

```diff
@@ -0,0 +1,104 @@
+/** @file
+  Shell command `log` to print the contents of the log buffer.
+
+  Copyright (c) 2018, Intel Corporation. All rights reserved.<BR>
+  This program and the accompanying materials
+  are licensed and made available under the terms and conditions of the BSD License
+  which accompanies this distribution.  The full text of the license may be found at
+  http://opensource.org/licenses/bsd-license.php.
+
+  THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+  WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+
+**/
+
+#include <Library/ShellLib.h>
+#include <Library/DebugLib.h>
+#include <Library/HobLib.h>
+#include <Guid/LoaderFspInfoGuid.h>
+#include <Library/ConsoleInLib.h>
+#include <Library/ConsoleOutLib.h>
+#include <Library/BootloaderCommonLib.h>
+
+/**
+  Print the contents of the log buffer
+
+  @param[in]  Shell        shell instance
+  @param[in]  Argc         number of command line arguments
+  @param[in]  Argv         command line arguments
+
+  @retval EFI_SUCCESS
+
+**/
+STATIC
+EFI_STATUS
+EFIAPI
+ShellCommandLogFunc (
+  IN SHELL  *Shell,
+  IN UINTN   Argc,
+  IN CHAR16 *Argv[]
+  );
+
+CONST SHELL_COMMAND ShellCommandLog = {
+  L"log",
+  L"Print messages stored in boot log buffer",
+  &ShellCommandLogFunc
+};
+
+/**
+  Print the contents of the log buffer
+
+  @param[in]  Shell        shell instance
+  @param[in]  Argc         number of command line arguments
+  @param[in]  Argv         command line arguments
+
+  @retval EFI_SUCCESS
+
+**/
+STATIC
+EFI_STATUS
+EFIAPI
+ShellCommandLogFunc (
+  IN SHELL  *Shell,
+  IN UINTN   Argc,
+  IN CHAR16 *Argv[]
+  )
+{
+  DEBUG_LOG_BUFFER_HEADER *LogBufHdr;
+  CONST UINTN              LinesPerPage = 30;
+  UINTN                    PageLineCount, Index;
+  UINT8                    Buf[1];
+  BOOLEAN                  Paged = FALSE;
+
+  for (Index = 1; Index < Argc; Index++) {
+    if (StrCmp (Argv[Index], L"-h") == 0) {
+      goto usage;
+    }
+    if (StrCmp (Argv[Index], L"-p") == 0) {
+      Paged = TRUE;
+    }
+  }
+
+  PageLineCount = 0;
+  LogBufHdr = (DEBUG_LOG_BUFFER_HEADER *) GetDebugLogBufferPtr ();
+  for (Index = 0; Index < (LogBufHdr->UsedLength - LogBufHdr->HeaderLength); Index++) {
+    ConsoleWrite ((UINT8 *)&LogBufHdr->Buffer[Index], 1);
+
+    // Page out the log contents if requested
+    if (Paged && (LogBufHdr->Buffer[Index] == '\n') && (++PageLineCount == LinesPerPage)) {
+      ShellPrint (L"[Press any key to continue...]");
+      ConsoleRead (Buf, 1);
```

Comment:
Could we support exiting the command by pressing a specific key (e.g. by pressing ESC key)? this way, we don't need press keys all the way to the end of log.

Reply:
Good idea. Added!